### PR TITLE
AP_Compass: Compare with string

### PR DESCRIPTION
**Pull request by muramura**
If you get a string from a device, determine it by the string.

### CODE REVIEW

**Review by peterbarker — CHANGES_REQUESTED** [review]
Saves a few bytes, which is a bit of a surprise:
```
Board,AP_Periph,blimp,bootloader,copter,heli,iofirmware,plane,rover,sub
Durandal,,0,*,-8,0,,0,-8,-8
HerePro,*,,,,,,,,
Hitec-Airspeed,*,,,,,,,,
KakuteH7-bdshot,,-8,*,-8,-8,,-8,-8,-8
MatekF405,,-8,*,-8,-8,,-8,-8,-8
Pixhawk1-1M-bdshot,,-8,,-8,-8,,-8,-8,-8
f103-QiotekPeriph,*,,,,,,,,
f303-Universal,-8,,,,,,,,
iomcu,,,,,,*,,,
revo-mini,,-8,*,-8,-8,,-8,-8,-8
skyviper-v2450,,,,*,,,,,
```

The `strncmp` isn't a natural fit for this data as they're not actually null-terminated strings.  The `strncmp` saves the situation, but still, we can do better.

Please change this to use `memcmp`.  `memcmp` can be *much* more efficient than `strncmp` and is more appropriate for this buffer.

File: libraries/AP_Compass/AP_Compass_HMC5843.cpp [hunk 1]
```diff
@@ -332,9 +332,7 @@ bool AP_Compass_HMC5843::_check_whoami()
         // can't talk on bus
         return false;        
     }
-    if (id[0] != 'H' ||
-        id[1] != '4' ||
-        id[2] != '3') {
+    if (strncmp((const char*)id, "H43", sizeof(id)) != 0) {
```
Comment (khancyr) [inline libraries/AP_Compass/AP_Compass_HMC5843.cpp:335]:
```suggestion
    if (memcmp(id, "H43", 3) != 0) {
```

**Review by khancyr — CHANGES_REQUESTED** [review]
memcmp is indeed the way to go in this one.

**Comment by peterbarker** [comment]
Replaced by https://github.com/ArduPilot/ardupilot/pull/28191

**Pull request closed without merging**